File: java/125.valid-palindrome.java
/*
 * @lc app=leetcode id=125 lang=java
 *
 * [125] Valid Palindrome
 */
class Solution {
    public boolean isPalindrome(String s) {
        if (s == null || "".equals(s)) return true;
        int i = 0, j = s.length() - 1;
        s = s.toLowerCase();
        while (i < j) {
            while (i < j && !isValid(s.charAt(i))) i++;
            while (i < j && !isValid(s.charAt(j))) j--;
            if (s.charAt(i++) != s.charAt(j--)) {
                System.out.println(s.charAt(i-1) + " " + s.charAt(j+1));
                return false;
            }
        }
        return true;
    }

    private boolean isValid(char c) {
       int v = c - 'a';
       if (0 <= v && v <= 25) return true;
       v =  c - '0';
       if (0 <= v && v <= 9) return true;
       return false;
    }

}

File: go/125.valid-palindrome.go
package goLeetCode

import "strings"

/*
 * @lc app=leetcode id=125 lang=golang
 *
 * [125] Valid Palindrome
 */

// @lc code=start

func isPalindrome(s string) bool {
	if len(s) == 0 {
		return true
	}
	const alpha = "abcdefghijklmnopqrstuvwxyz0123456789"
	i, j := 0, len(s)-1
	for i < j {
		lo := strings.ToLower(string(s[i]))
		if !strings.Contains(alpha, lo) {
			i++
			continue
		}
		hi := strings.ToLower(string(s[j]))
		if !strings.Contains(alpha, hi) {
			j--
			continue
		}
		if lo != hi {
			return false
		} else {
			i++
			j--
		}
	}
	return true
}

// @lc code=end
